File: src/BackEnd/go-services/WalletApi/internal/models/api_response.go
package models

import "math"

type ApiResponse[T any] struct {
	Status     int            `json:"status"`
	Message    string         `json:"message"`
	Data       interface{}    `json:"data,omitempty"`
	Pagination *Pagination[T] `json:"pagination,omitempty"`
	Error      *Error         `json:"error"`
}

type Error struct {
	Type     string `json:"type"`
	Title    string `json:"title"`
	Status   int    `json:"status"`
	Detail   string `json:"detail,omitempty"`
	Instance string `json:"instance,omitempty"`
}

type Pagination[T any] struct {
	Page       int `json:"page"`
	PerPage    int `json:"per_page"`
	PagesTotal int `json:"pages_total"`
	TotalItems int `json:"total_items"`
	Data       []T `json:"-"`
}

func PaginateData[T any](items []T, totalItems, page, perPage int) Pagination[T] {
	if page < 1 {
		page = 1
	}
	if perPage < 1 {
		perPage = 10
	}

	totalPages := int(math.Ceil(float64(totalItems) / float64(perPage)))
	if totalPages == 0 {
		totalPages = 1
	}

	return Pagination[T]{
		Page:       page,
		PerPage:    perPage,
		PagesTotal: totalPages,
		TotalItems: totalItems,
		Data:       items,
	}
}

File: src/BackEnd/dotnet-services/IdentityManagementApi/src/AdvertisingSystem.Identity.Api/ViewModels/ApiResponse.cs
using System.Net;

namespace AdvertisingSystem.Identity.Api.ViewModels;

public class ApiResponse<T> where T : class
{
    public ApiResponse(HttpStatusCode status)
    {
        Status = status;
    }

    public ApiResponse(HttpStatusCode status, string message)
    {
        Status = status;
        Message = message;
    }

    public ApiResponse(HttpStatusCode status, string message, T data)
    {
        Status = status;
        Message = message;
        Data = data;
    }

    public ApiResponse(T data)
    {
        Data = data;
        Status = HttpStatusCode.OK;
        Message = "The operation was successful";
    }

    public ApiResponse(HttpStatusCode status, string message, T data, Pagination? pagination, ApiError? error)
    {
        Status = status;
        Message = message;
        Data = data;
        Pagination = pagination;
        Error = error;
    }

    public HttpStatusCode Status { get; set; }
    public string Message { get; set; }
    public T Data { get; set; }
    public Pagination? Pagination { get; set; }
    public ApiError? Error { get; set; }
}

public struct ApiError
{
    public string Type { get; set; }
    public string Title { get; set; }
    public HttpStatusCode Status { get; set; }
    public string Detail { get; set; }
    public string Instance { get; set; }
}

public struct Pagination
{
    public int Page { get; set; }
    public int PerPage { get; set; }
    public int PagesTotal { get; set; }
    public int TotalItems { get; set; }
}